### ISSUE TITLE
"mouse_entered" and "mouse_exited" signals are broken while holding down a mouse button

### ISSUE THREAD

**Issue by PostPollux** [godotengine/godot#20881]
**Godot version:**
fbb5ca4

**OS/device including version:**
Linux Manjaro

**Issue description:**
The "mouse_entered" and "mouse_exited" signals of control nodes don't get fired while holding any mouse button down when the click (any mouse button down) started on a control that does not ignore the mouse. 

it's actually because in viewport.cpp in line 1843 to 1850 "over" get's set to the mouse_focus
```
if (gui.mouse_focus) {
	over = gui.mouse_focus;
	//recompute focus_inv_xform again here
} else {
	over = _gui_find_control(mpos);
}
```

and shortly after (  line 1884 to 1893 ) the mouse_exited and mouse_entered signals get not fired because the gui.mouse_focus and the gui.mouse_over are the same

```
if (over != gui.mouse_over) {

	if (gui.mouse_over)
		gui.mouse_over->notification(Control::NOTIFICATION_MOUSE_EXIT);

	_gui_cancel_tooltip();

	if (over)
		over->notification(Control::NOTIFICATION_MOUSE_ENTER);
}
```

This may be a desired behavior, for a lot of cases.
But there are also cases where it is very useful to send those signals. (I'll update this post soon to show such a case)

**Steps to reproduce:**
1. Connect the "mouse_entered" or "mouse_exited" signal of any control node
2. add another control node, that does not ignore the mouse
3. start your drag on the control element of point 2 and drag over the control element from point 1
4. notice that no "mouse_entered" and no "mouse_exited" signals get fired

**Comment by spongeboburu** [comment]
Confirmed on my master build too. Note: if you set mouse filter to "ignore" on the parent control the color rectangle signals works as expected.

**Comment by jeremyz** [comment]
linked to #23296

**Comment by groud** [comment]
Note from #32727, this does also happen for is_hovered()

**Comment by yuyubibibobibo** [comment]
> **Godot版本：**
> [fbb5ca4](https://github.com/godotengine/godot/commit/fbb5ca4d9733426aa3b161c7209942aa2bb67953)
> 
> **操作系统/设备，包括版本：**
> Linux Manjaro
> 
> **问题描述：**
> 当在不忽略鼠标的控件上单击（按下任何鼠标按钮）时，按住任意鼠标按钮时，不会触发控制节点的“ mouse_entered”和“ mouse_exited”信号。
> 
> 实际上是因为在1843至1850行的viewport.cpp中，“ over”被设置为mouse_focus
> 
> ```
> if (gui.mouse_focus) {
> 	over = gui.mouse_focus;
> 	//recompute focus_inv_xform again here
> } else {
> 	over = _gui_find_control(mpos);
> }
> ```
> 
> 并且在之后（第1884至1893行）不久，由于gui.mouse_focus和gui.mouse_over相同，因此未触发mouse_exited和mouse_entered信号
> 
> ```
> if (over != gui.mouse_over) {
> 
> 	if (gui.mouse_over)
> 		gui.mouse_over->notification(Control::NOTIFICATION_MOUSE_EXIT);
> 
> 	_gui_cancel_tooltip();
> 
> 	if (over)
> 		over->notification(Control::NOTIFICATION_MOUSE_ENTER);
> }
> ```
> 
> 在很多情况下，这可能是理想的行为。
> 但是在某些情况下，发送这些信号非常有用。（我将尽快更新此帖子以显示这种情况）
> 
> **重现步骤：**
> 
> 1. 连接任何控制节点的“ mouse_entered”或“ mouse_exited”信号
> 2. 添加另一个不忽略鼠标的控制节点
> 3. 在点2的控件元素上开始拖动，然后从点1的控件元素上拖动
> 4. 请注意，没有触发“ mouse_entered”和“ mouse_exited”信号

how solve it...cry...＃35262

**Comment by swenner** [comment]
Still an issue in Godot 3.2.2beta ed27b7e6b973eb6be6fdac4439e99a1907b9cb58

**Comment by MadDogMayCry0** [comment]
Still in 3.2.4RC2.
Try to click on bottom element. With out relesing it move cursor away from element!
[New Game Project.zip](https://github.com/godotengine/godot/files/6037475/New.Game.Project.zip)

**Comment by MrJoshBowman** [comment]
managed to get around this by making my own _input(event:InputEventMouseButton) function.
I then had to make sure these two lines were executed as part of the code:
```
event.set_pressed(false)
set("mouse_filter", Control.MOUSE_FILTER_STOP)
```
This only worked with an _input() function, not a _gui_input() function

**Comment by MadDogMayCry0** [comment]
> удалось обойти это, создав мою собственную функцию _input (event: InputEventMouseButton).
> Затем мне нужно было убедиться, что эти две строки выполняются как часть кода при нажатии мыши:
> 
> ```
> event.set_pressed(false)
> set("mouse_filter", Control.MOUSE_FILTER_STOP)
> ```
> 
> Затем я сбрасываю mouse_filter на ..._ PASS при отпускании мыши.
> Это работало только с функцией _input (), а не с функцией _gui_input ()

Could you show full code pls?

**Comment by MadDogMayCry0** [comment]
Ok, take a look an attachment, it's works well :)
[Mouse pressed signals bug.zip](https://github.com/godotengine/godot/files/6335217/Mouse.pressed.signals.bug.zip)

**Comment by Banderi** [comment]
> FIX :)
> [Mouse pressed signals bug.zip](https://github.com/godotengine/godot/files/6343048/Mouse.pressed.signals.bug.zip)

Can not get this to work. The line `get_global_rect().has_point(get_global_mouse_position())` does not take into account either scale or rotation of the node :(

**Comment by Banderi** [comment]
> > > FIX :)
> > > [Mouse pressed signals bug.zip](https://github.com/godotengine/godot/files/6343048/Mouse.pressed.signals.bug.zip)
> > 
> > 
> > Can not get this to work. The line `get_global_rect().has_point(get_global_mouse_position())` does not take into account either scale or rotation of the node :(
> 
> ```gdscript
> get_global_rect().has_point(get_local_mouse_position())
> ```

You literally just reposted the code line I mentioned... As I said, it doesn't work.

**Comment by MadDogMayCry0** [comment]
@Banderi Local mouse position, rotating, scaling
[Mouse pressed signals bug.zip](https://github.com/godotengine/godot/files/6483692/Mouse.pressed.signals.bug.zip)

**Comment by deanl42** [comment]
Wasted hours thinking it was my code :-( v3.3.1
Appreciate the work around but hoping this gets a fix/resolution

**Comment by Zylann** [comment]
This is probably related to a common UI behavior called "mouse capture". The idea is to make it easier to implement dragging logic, by making the widget receive mous events and appear pressed as long as the mouse was pressed inside and remains pressed even if it exits the area of the control.

For example, take a scrollbar. Press it, keep the mouse held, and move away from it. It will remain pressed, and will still respond to your movements, while every other control will ignore the mouse. You can reproduce this behavior on non-Godot UIs as well, even on controls that are not usually draggable (see how buttons remain pressed when you do this too). Things can differ a bit though, sometimes hover effects don't follow this, while pressed effects do.

About `mouse_entered` and `mouse_exited`: if these were changed to always fire when the mouse goes out of the control's area, regardless of what's actually happening, then likely a bunch of existing code might need to be fixed somehow, without being able to rely on them. Or maybe controls can have an option to turn off mouse capture.

**Comment by deanl42** [comment]
> This is probably related to a common UI behavior, which I saw called "mouse capture" on Windows. The idea is to make it easier to implement exclusive dragging logic, by making the control receive mouse events and appear pressed as long as the mouse was pressed inside and remains pressed even if it exits the area of the control. I would not be surprised if this wasnt documented because it's a subtle behavior happening with many widgets, and has no associated option.
> 
> For example, take a scrollbar. Press it, keep the mouse held, and move away from it. It will remain pressed, and will still respond to your movements, while every other control will ignore the mouse. You can reproduce this behavior on non-Godot UIs as well, even sometimes on controls that are not usually draggable (see how some buttons remain pressed when you do this too). Things can differ a bit though, because there are many UI systems around. Sometimes hover effects don't follow this, while pressed effects do. Godot happens to do both, including enter/exit.
> 
> About `mouse_entered` and `mouse_exited`: arguably, many widgets probably don't need mouse capture. But if these were changed to always fire when the mouse goes in and out of the control's area, regardless of what's actually happening, then likely a bunch of existing code might need to be fixed somehow, without being able to rely on them. Or maybe controls can have an option to turn off mouse capture.

Thanks for that, makes sense.  If this is the case then I would def upvote an option to switch it off for certain scenarios.

**Comment by Calinou** [comment]
Out of curiosity, does this behavior happen with CollisionObject2D/3D's `mouse_enter`/`mouse_exit` signals too?

**Comment by golddotasksquestions** [comment]
This is also a huge problem for `gui_input(event)` signals from any Control based node.

For example when trying to trigger an animation when the left-mouse is released over the Control:

```
func _on_Control_gui_input(event):
	if event.is_action_released("LMB"):
		$AnimationPlayer.play("New Anim")
```

This should trigger the animation only when the left mouse is released over the Control, but if clicked onto the Control then dragged the mouse away from the control (while still being pressed) the animation will play, regardless of where the mouse is  released.

This seems to be connected to the mouse_enter and mouse_exit signal, as can be easily tested with this:

```
var mouse_hover = false

func _on_Control_gui_input(event):
	if event.is_action_released("LMB") and mouse_hover:
		$AnimationPlayer.play("New Anim")
		print("gui ",mouse_hover)

func _on_Control_mouse_entered():
	mouse_hover = true
	print("mouse_hover = true")

func _on_Control_mouse_exited():
	mouse_hover = false
	print("mouse_hover = false")
```

Both "gui True" and "mouse_hover = false" will print only after mouse release.

**Comment by MadDogMayCry0** [comment]
> 
> 
> This is also a huge problem for `gui_input(event)` signals from any Control based nodes.
> 
> For example when trying to trigger an animation when the left-mouse is released over the Control:
> 
> ```
> func _on_Control_gui_input(event):
> 	if event.is_action_released("LMB"):
> 		$AnimationPlayer.play("New Anim")
> ```
> 
> This should trigger the animation only when the left mouse is released over the Control, but if clicked onto the Control then dragged the mouse away from the control (while still being pressed) the animation will play, regardless of where the mouse is released.
> 
> This seems to be connected to the mouse_enter and mouse_exit signal, as can be easily tested with this:
> 
> ```
> var mouse_hover = false
> 
> func _on_Control_gui_input(event):
> 	if event.is_action_released("LMB") and mouse_hover:
> 		$AnimationPlayer.play("New Anim")
> 		print("gui ",mouse_hover)
> 
> func _on_Control_mouse_entered():
> 	mouse_hover = true
> 	print("mouse_hover = true")
> 
> func _on_Control_mouse_exited():
> 	mouse_hover = false
> 	print("mouse_hover = false")
> ```
> 
> Both "gui True" and "mouse_hover = false" will print only after mouse release.

https://github.com/godotengine/godot/issues/20881#issuecomment-841614529

**Comment by golddotasksquestions** [comment]
@MadDogMayCry0 
I can't get your workaround to work with my gui_input(event) example. I suppose it would be possible if I copy it line by line, use _input(event) instead of gui_input(event) and separate scripts like you did, but this is way too complicated for what I want to do.

**Comment by MadDogMayCry0** [comment]
> 
> 
> @MadDogMayCry0 I can't get your workaround to work with my gui_input(event) example. I suppose it would be possible if I copy it line by line, use _input(event) instead of gui_input(event) and separate scripts like you did, but this is way too complicated for what I want to do.

> 
> 
> @MadDogMayCry0 even though it works, what you do with `if event.is_pressed(): event.pressed = false` is to by pass the code that starts here : https://github.com/godotengine/godot/blob/3.x/scene/main/viewport.cpp#L1861 and thus break dragging modal windows and some yet unclear to me gui focus related code.


Natürlich müssen Sie meine Logik an das anpassen, was Sie mit Ihnen vorhaben. Meine Logik erklärt und demonstriert nur, wie dieses Problem umgangen wird.

**Comment by jeremyz** [comment]
> 
see my 2 pull requests #54565 and #54552 and also #67791 from @Sauermann

**Comment by ondesic** [comment]
Can confirm this problem in 3.5.2 Stable

**Comment by KoBeWi** [comment]
Can't reproduce in 4.1.
Can someone confirm?

**Comment by Sauermann** [comment]
Tested in v4.2.dev.custom_build [c3b0a92c3].
I can't reproduce the problem in the simple case, that starting Control and hovering Control are in the same Viewport. That seems to be fixed.

However, when starting Control and hovering Control are in different Viewports (SubViewport or Window), then I can reproduce the problem in many cases.
MRP for this is available in: https://github.com/godotengine/godot/pull/67791#issue-1419768978

**Comment by KoBeWi** [comment]
Closing then. The more complex case is a separate issue.

**Comment by Sauermann** [comment]
I didn't test it, but It is likely that this got fixed by #59555.

**Comment by kdserra** [comment]
Experiencing this on Godot 3.5.2

I am using using this to feature for drag & dropping items in my inventory system.

Move move over item slot -> Triggers MOUSE ENTERED (Good)
Hold down LMB -> Triggers DOWN (Good)
Move mouse outside item slot -> Triggers Nothing (Bad) Expected: MOUSE EXITED
Let go of LMB -> Triggers UP, and MOUSE EXITED (Half correct)

```
        public void _on_Button_button_down()
        {
            GD.Print("DOWN");
        }

        public void _on_Button_button_up()
        {
            GD.Print("UP");
        }

        public void _on_Button_mouse_entered()
        {
            GD.Print("MOUSE ENTERED");
        }

        public void _on_Button_mouse_exited()
        {
            GD.Print("MOUSE EXITED");
        }
```

![Untitled](https://github.com/godotengine/godot/assets/25918580/8dd36ed4-d969-4159-bf5b-466a641ebf3f)

**Comment by KoBeWi** [comment]
The issue is only fixed in Godot 4.1.

**Comment by NicolasVidal** [comment]
Does anyone know if it will be fixed in 3.5.x ?

**Comment by josepvalls** [comment]
Just hit this issue. Doesn't look fixed in 3.5.3 stable. Bummer.

**Comment by Poobslag** [comment]
Can confirm this bug is still not fixed in 3.5.3 stable.
